### PR TITLE
New option to always copy over key to SSL object

### PR DESCRIPTION
**Pull request by ColtonWilley**
# Description

Initial implementation of new option to always copy over key to SSL ctx

Fixes zd 18662

### CODE REVIEW

File: wolfssl/wolfcrypt/settings.h [hunk 1]
```diff
@@ -3581,6 +3581,11 @@ extern void uITRON4_free(void *p) ;
     #define WOLFSSL_COPY_CERT
 #endif
 
+#if defined(OPENSSL_ALL) && !defined(WOLFSSL_NO_COPY_KEY)
+    #undef WOLFSSL_COPY_KEY
```
Comment (dgarske) [inline wolfssl/wolfcrypt/settings.h:3585]: Please add a code comment for purpose of `WOLFSSL_COPY_KEY` and use-case it resolves.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfssl/wolfcrypt/settings.h (hunk 1 above)
Comment (ColtonWilley) [reply wolfssl/wolfcrypt/settings.h:3591]: Comments have been added.

**Review by dgarske — APPROVED** [review]

File: src/internal.c [hunk 2]
```diff
@@ -6829,7 +6829,22 @@ int SetSSL_CTX(WOLFSSL* ssl, WOLFSSL_CTX* ctx, int writeDup)
     ssl->buffers.certChainCnt = ctx->certChainCnt;
 #endif
 #ifndef WOLFSSL_BLIND_PRIVATE_KEY
+#ifdef WOLFSSL_COPY_KEY
+    if (ctx->privateKey != NULL) {
+        if (ssl->buffers.key != NULL) {
+            FreeDer(&ssl->buffers.key);
```
Comment (JacobBarthelmeh) [inline src/internal.c:6835]: I think if free'ing a previous key here we need a check for ssl->buffers.weOwnKey and follow the same pattern as wolfSSL_UnloadCertsKeys().

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 2 above)
Comment (ColtonWilley) [reply src/internal.c:6835]: I actually disagree. Unlike wolfSSL_UnloadCertsKeys() we are with certainty going to set the ssl->buffers.key. Checking on weOwnCert as well just introduces the chance that through mismanagement if weOwnCert is incorrectly not set at the time of this check and we leak memory. This flag is also partially to ensure behavior replication for openssl compat. Regardless of the state of weOwnCert, if this flag is enabled we should always be copying the key material in my opinion as that is what openssl does.

Comment (JacobBarthelmeh) [reply src/internal.c:6835]: Biggest concern was a case state where the pointer is pointing to the key from the WOLFSSL_CTX. I don't think that state could be hit here though when looking more. We do not check before assigning for the blinding case.

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**